### DISCUSSION TITLE
Batch integration onto already integrated seurat object

### DISCUSSION THREAD

**Discussion by MiTPenguin** [satijalab/seurat#5681] [Q&A]
*reposting it here, since maybe this is not an "issue"?*

Hello, I am working on a set of immune cell data, generated using droplet-based protocol. Previously, I followed the SCT + CCA batch correction vignette, and got a pretty well corrected data set, where known phenotypes are pretty well separated, but there aren't too many sample-specific clusters. Seems like an okay trade-off between preserving interesting biology and removing batch effects.

So this resulting seurat object has an "integrated" assay (also an RNA and an SCT assay) that's active by default. I have collected additional data that I want to integrate into this existing seurat object, but I'm not finding a lot of info on the correct way to do it.

Re-integrate everything with CCA is going to take too much time and memory, so I was hoping to use the existing integrated seurat object as a reference, and apply rpca to integrate the new data onto it.

I am unclear on the details of how to implement this, though. I assume I have to SCT + PCA process each of the new data set separately, and them in a list along with the already integrated seurat object. I assume, since I want to integrate onto the PCA space of an integrated dataset, I have to keep the active assay of the already integrated seurat object as "integrated," while other new datasets would have active assay of "SCT". Is this correct? Then do I still have to run PrepSCTIntegration?

When I tried to do this, the resulting integrated object ends up with a bunch of NaN in the data slot, corresponding to the number of cells added in the new unintegrated dataset. The genes exist in all of the data sets that I'm using.

Any help would be greatly appreciated!

**Comment by DiracZhu1998** [comment]
For integrating a large dataset containing multiple replicates (lets say 100 replicates), I usually integrate 5 replicate to build a low-level integrated dataset, then I integrate "these already integrated dataset". 
Because it's not only time consuming but also having terrible integration effect in directly integrating a large number of replicates.
I can share my script here.


suppressMessages(suppressWarnings(library(Seurat)))
suppressMessages(suppressWarnings(library(ggplot2)))
suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(cowplot)))
suppressMessages(suppressWarnings(library(plyr)))

## ---------------------------
##
## Script name:
##
## Purpose of script: Integration (SCTransform + CCA + MNN) and clustering based on given parameters
##
## Author: Yuanzhen Zhu, Jiahao Xiang
##
## Date Created: 2020-11-15, Data Modified: 2021-10-08, 2021-12-21, 2022-03-25
##
## Version: V2.3
##
## ---------------------------
##
## Notes: This script can be universal and general used!
##
## ---------------------------


parser = argparse::ArgumentParser(description = "Script to integrate scRNA data")
parser$add_argument('-I', '--input', help = 'Input file containing path of RDS file needing integration')
parser$add_argument('-F', '--nf', help = 'Number of HVGs')
parser$add_argument('-TN', '--tn', help = 'top markers plot heatmap')
parser$add_argument('-D', '--dim',help = 'PCA dim usage')
parser$add_argument('-R', '--res',help = 'Map resolution usage')
parser$add_argument('-O', '--out',help = 'out directory')
args = parser$parse_args()

nf.usage <- as.numeric(if(!is.null(args$nf)) args$nf else 3000)
dim.usage <- as.numeric(if(!is.null(args$dim)) args$dim else 30)
res.usage <- as.numeric(if(!is.null(args$res)) args$res else 0.6)
tn.usage <- as.numeric(if(!is.null(args$tn)) args$tn else 20)


# read input RDS, and for loop it
path.list <- read.table(args$input)
id.list <- c()
for (i in 1:nrow(path.list)){
    abc = strsplit(path.list[i,1], split = "/")
    id <- abc[[1]][lengths(abc)]
    del_id <- path.list[i,2]
    assign(del_id, readRDS(path.list[i,1]))
    id.list <- c(id.list, del_id)
}

# build list containing objects
rds.list <- lapply(id.list, get)

rds.list
# Normalization and FindHVGs

rds.list <- lapply(X = rds.list, FUN = SCTransform)
features <- SelectIntegrationFeatures(object.list = rds.list, nfeatures = nf.usage)

print(rds.list)

# Integration
rds.list <- PrepSCTIntegration(object.list = rds.list, anchor.features = features)
warnings()
Combined.anchors <- FindIntegrationAnchors(object.list = rds.list, dims = 1:dim.usage, anchor.features = features, normalization.method = "SCT", reduction = "cca")
Combined <- IntegrateData(anchorset = Combined.anchors, normalization.method = "SCT", dims = 1:dim.usage)
DefaultAssay(Combined) <- "integrated"

# pca then heatmap
Combined <- RunPCA(Combined, npcs = dim.usage)
pdf(paste0(args$out,"/SCT_CCA_MNN_DimHeatmap.pdf"), width = 30, height = 20)
DimHeatmap(Combined, dims = 1:dim.usage, cells = 100, balanced = TRUE)
dev.off()

#Combined <- JackStraw(Combined, num.replicate = 50)
#Combined <- ScoreJackStraw(Combined, dims = 1:20)
#pdf(paste0(args$out, "/SCT_CCA_MNN_JackStrawPlot.pdf"))
#JackStrawPlot(Combined, dims = 1:20)
#dev.off()
#pdf(paste0(args$out,"/SCT_CCA_MNN_ElbowPlot.pdf"))
#ElbowPlot(Combined)
#dev.off()

Combined <- RunTSNE(Combined, dims = 1:(dim.usage-5), reduction = "pca", check_duplicates = FALSE)
Combined <- RunUMAP(Combined, dims = 1:(dim.usage-5), reduction = "pca")
Combined <- FindNeighbors(Combined, reduction = "pca", dims = 1:(dim.usage-5))
Combined <- FindClusters(Combined, resolution = res.usage, n.start=10)


DefaultAssay(Combined) <- "RNA"
p1 <- DimPlot(Combined, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(Combined, reduction = "umap", label = T, pt.size=0.5)
p3 <- DimPlot(Combined, reduction = "tsne", group.by = "orig.ident")
p4 <- DimPlot(Combined, reduction = "tsne", label = T, pt.size=0.5)

pdf(paste0(args$out, "/SCT_CCA_MNN_cell_cluster.umap.pdf"), height=10, width=20)
plot_grid(p1, p2)
dev.off()
pdf(paste0(args$out, "/SCT_CCA_MNN_cell_cluster.tsne.pdf"), height=10, width=20)
plot_grid(p3, p4)
dev.off()

pdf(paste0(args$out, "/SCT_CCA_MNN_cell_cluster_umi.pdf"), height=10, width=20)
p5 <- ggplot(as.data.frame(Combined@reductions$umap@cell.embeddings), aes(x=UMAP_1, y=UMAP_2, color=Combined@meta.data$nCount_RNA)) + geom_point(size=0.5) + theme(legend.position = "none") + theme_bw() + labs(color="UMIs") + scale_color_viridis_c(direction = -1)
p6 <- ggplot(as.data.frame(Combined@reductions$tsne@cell.embeddings), aes(x=tSNE_1, y=tSNE_2, color=Combined@meta.data$nCount_RNA)) + geom_point(size=0.5) + theme(legend.position = "none") + theme_bw() + labs(color="UMIs") + scale_color_viridis_c(direction = -1)
plot_grid(p5, p6)
dev.off()

pdf(paste0(args$out,"/SCT_CCA_MNN_cell_cluster_gene.pdf"),height=10,width=20)
p7 <- ggplot(as.data.frame(Combined@reductions$umap@cell.embeddings), aes(x=UMAP_1, y=UMAP_2, color=Combined@meta.data$nFeature_RNA)) + geom_point(size=0.5) + theme(legend.position = "none") + theme_bw() + labs(color="genes") + scale_color_viridis_c(direction = -1)
p8 <- ggplot(as.data.frame(Combined@reductions$tsne@cell.embeddings), aes(x=tSNE_1, y=tSNE_2, color=Combined@meta.data$nFeature_RNA)) + geom_point(size=0.5) + theme(legend.position = "none") + theme_bw() + labs(color="genes") + scale_color_viridis_c(direction = -1)
plot_grid(p7, p8)
dev.off()

markers <- FindAllMarkers(Combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
write.table(markers,paste0(args$out,'/SCT_CCA_MNN_markers.txt'),sep = '\t',row.names =FALSE, col.names =TRUE,quote =FALSE)
top <- markers %>% group_by(cluster) %>% top_n(n = tn.usage, wt = avg_log2FC)

####### Pay attention! ####
the input file contains:
/Users/xxx/path/low_level_integrated_data.1.RDS label
/Users/xxx/path/low_level_integrated_data.2.RDS label
...

**Reply by qdong2023** [reply]
Hi, @DiracZhu1998, I wonder if I should use the "RNA" assay or "integrated" assay when integrating new datasets with the ones that have been already integrated? Thanks!

**Comment by qdong2023** [comment]
Hi, @DiracZhu1998, I wonder if I should use the "RNA" assay or "integrated" assay when integrating new datasets with the ones that have been already integrated? Thanks!

**Reply by DiracZhu1998** [reply]
It depends, if you are trying to integrate a few samples/conditions, you should use "RNA assay". If you have a lot of bio and tech replicates (for example, 10 conditions with 5 technical replicates), you could first generate "condition" level atlases, then build the whole atlas based on the "condition" level atlas's "integrated assay". 
The idea behind it is Seurat cannot handle complex experimental conditions like harmony. You could try harmony also, It usually works better on complex-designed scRNA integration.

**Reply by qdong2023** [reply]
Thanks for your reply! I have tried the harmony, however, it looks like over integrated. Another related question is whether the integration is just re-clustering/grouping cells from multiple datasets? Does it also correct the gene expression across batches? For example, I have two genotypes and 15 samples for each genotype from 2 batches, gene A in the same genotype may be differently expressed across batches, so can the Seurat integration method correct gene expression level? Thanks!
